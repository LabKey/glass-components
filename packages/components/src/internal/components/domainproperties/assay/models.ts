/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { List, Record as ImmutableRecord } from 'immutable';

import { getServerContext, Utils } from '@labkey/api';

import { DomainDesign, FieldErrors } from '../models';
import { AppURL } from '../../../url/AppURL';
import { getAppHomeFolderPath } from '../../../app/utils';
import { Container } from '../../base/models/Container';

// See ExpProtocol.Status in 'platform' repository.
export enum Status {
    Active = 'Active',
    Archived = 'Archived',
}

export interface ProtocolTransformScript {
    runOnEdit: boolean;
    runOnImport: boolean;
    scriptPath: string;
}

export class AssayProtocolModel extends ImmutableRecord({
    allowBackgroundUpload: false,
    allowEditableResults: false,
    allowQCStates: false,
    allowTransformationScript: false,
    allowPlateMetadata: false,
    autoCopyTargetContainer: undefined,
    autoCopyTargetContainerId: undefined,
    autoLinkCategory: undefined,
    availableDetectionMethods: undefined,
    availableMetadataInputFormats: undefined,
    availablePlateTemplates: undefined,
    backgroundUpload: false,
    description: undefined,
    domains: undefined,
    editableResults: false,
    editableRuns: false,
    exception: undefined,
    metadataInputFormatHelp: undefined,
    moduleTransformScripts: undefined,
    name: undefined,
    plateMetadata: undefined,
    protocolId: undefined,
    protocolParameters: undefined,
    protocolTransformScripts: undefined,
    providerName: undefined,
    saveScriptFiles: false,
    selectedDetectionMethod: undefined,
    selectedMetadataInputFormat: undefined,
    selectedPlateTemplate: undefined,
    status: Status.Active,
    qcEnabled: undefined,
    excludedContainerIds: undefined,
}) {
    declare allowBackgroundUpload: boolean;
    declare allowEditableResults: boolean;
    declare allowQCStates: boolean;
    declare allowTransformationScript: boolean;
    declare allowPlateMetadata: boolean;
    declare autoCopyTargetContainer: {};
    declare autoCopyTargetContainerId: string;
    declare autoLinkCategory: string;
    declare availableDetectionMethods: [];
    declare availableMetadataInputFormats: {};
    declare availablePlateTemplates: [];
    declare backgroundUpload: boolean;
    declare description: string;
    declare domains: List<DomainDesign>;
    declare editableResults: boolean;
    declare editableRuns: boolean;
    declare exception: string;
    declare metadataInputFormatHelp: any;
    declare moduleTransformScripts: List<string>;
    declare name: string;
    declare plateMetadata: boolean;
    declare protocolId: number;
    declare protocolParameters: any;
    declare protocolTransformScripts: List<ProtocolTransformScript>;
    declare providerName: string;
    declare saveScriptFiles: boolean;
    declare selectedDetectionMethod: string;
    declare selectedMetadataInputFormat: string;
    declare selectedPlateTemplate: string;
    declare status: Status;
    declare qcEnabled: boolean;
    declare excludedContainerIds: string[];

    static create(raw: any): AssayProtocolModel {
        let domains = raw.domains || List<DomainDesign>();
        if (raw.domains && Utils.isArray(raw.domains)) {
            domains = List<DomainDesign>(
                raw.domains.map(domain => {
                    return DomainDesign.create(domain);
                })
            );
        }

        if (raw.protocolTransformScripts && Utils.isArray(raw.protocolTransformScripts)) {
            raw.protocolTransformScripts = List(raw.protocolTransformScripts);
        }
        if (raw.moduleTransformScripts && Utils.isArray(raw.moduleTransformScripts)) {
            raw.moduleTransformScripts = List<string>(raw.moduleTransformScripts);
        }

        // if this is not an existing assay, clear the name property so the user must set it
        const name = !raw.protocolId ? undefined : raw.name;

        // Issue 38685: for new assays, pre-select some required assay properties
        const model = new AssayProtocolModel({ ...raw, name, domains });
        if (model.isNew()) {
            if (model.allowDetectionMethodSelection() && model.availableDetectionMethods.length > 0) {
                raw.selectedDetectionMethod = List<string>(model.availableDetectionMethods).get(0);
            }
            if (model.allowPlateTemplateSelection() && model.availablePlateTemplates.length > 0) {
                raw.selectedPlateTemplate = List<string>(model.availablePlateTemplates).get(0);
            }
        }

        return new AssayProtocolModel({ ...raw, name, domains });
    }

    static serialize(model: AssayProtocolModel): any {
        // need to serialize the DomainDesign objects to remove the unrecognized fields
        const domains = model.domains.map(domain => {
            return DomainDesign.serialize(domain);
        });

        const json = model.merge({ domains }).toJS();

        // only need to serialize the id and not the autoCopyTargetContainer object
        delete json.autoCopyTargetContainer;
        delete json.exception;
        return json;
    }

    isActive(): boolean {
        return this.status === Status.Active;
    }

    getDomainByNameSuffix(name: string): DomainDesign {
        if (this.domains.size > 0) {
            return this.domains.find(domain => {
                return domain.isNameSuffixMatch(name);
            });
        }
    }

    get container(): string {
        const container = new Container(getServerContext().container);
        const domainContainerId = this.domainContainerId;

        return this.isNew()
            ? getAppHomeFolderPath(container)
            : domainContainerId === container.id
              ? container.path
              : domainContainerId;
    }

    get domainContainerId(): string {
        return this.getIn(['domains', 0, 'container']);
    }

    isNew(): boolean {
        return !this.protocolId;
    }

    allowPlateTemplateSelection(): boolean {
        return this.availablePlateTemplates && Utils.isArray(this.availablePlateTemplates);
    }

    allowDetectionMethodSelection(): boolean {
        return this.availableDetectionMethods && Utils.isArray(this.availableDetectionMethods);
    }

    allowMetadataInputFormatSelection(): boolean {
        return (
            this.availableMetadataInputFormats &&
            Utils.isObject(this.availableMetadataInputFormats) &&
            !Utils.isEmptyObj(this.availableMetadataInputFormats)
        );
    }

    validateTransformScripts(): string {
        if (this.protocolTransformScripts === undefined || this.protocolTransformScripts.size === 0) {
            return undefined;
        }

        // make sure we don't have any script inputs that are empty strings
        const hasEmptyScript = this.protocolTransformScripts.some(
            config => config?.scriptPath === undefined || config.scriptPath === null || config.scriptPath.length === 0
        );
        if (hasEmptyScript) {
            return 'Missing required transform script path.';
        }
    }

    getFirstDomainFieldError(): FieldErrors {
        const firstErrantDomain = this.domains.find(domain => domain.hasInvalidFields());
        return firstErrantDomain !== undefined ? firstErrantDomain.getInvalidFields().first().getErrors() : undefined;
    }

    isValid(): boolean {
        return this.hasValidProperties() && this.getFirstDomainFieldError() === undefined;
    }

    isGPAT(): boolean {
        return this.providerName.toLowerCase() === 'general';
    }

    hasValidProperties(): boolean {
        return (
            this.name !== undefined &&
            this.name !== null &&
            this.name.trim().length > 0 &&
            (!this.allowMetadataInputFormatSelection() || Utils.isString(this.selectedMetadataInputFormat)) &&
            (!this.allowDetectionMethodSelection() || Utils.isString(this.selectedDetectionMethod)) &&
            (!this.allowPlateTemplateSelection() || Utils.isString(this.selectedPlateTemplate)) &&
            this.validateTransformScripts() === undefined
        );
    }

    get hasBatchFields(): boolean {
        return this.getDomainByNameSuffix('Batch')?.fields.size > 0;
    }

    getUrl(): AppURL {
        return AppURL.create('assays', this.providerName, this.name);
    }
}
