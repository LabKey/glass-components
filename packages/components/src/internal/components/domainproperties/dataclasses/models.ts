/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Draft, immerable, produce } from 'immer';
import { Map, fromJS, OrderedMap, Set } from 'immutable';

import { DomainDesign, IDomainField, SystemField } from '../models';

import { IParentAlias } from '../../entities/models';

import { DATACLASS_DOMAIN_SYSTEM_FIELDS, SOURCE_DOMAIN_SYSTEM_FIELDS } from './constants';
import {getDuplicateAlias, parentAliasInvalid} from "../utils";

interface DataClassOptionsConfig {
    category: string;
    description: string;
    importAliases?: Map<string, string>;
    name: string;
    nameExpression: string;
    parentAliases?: OrderedMap<string, IParentAlias>;
    rowId: number;
    sampleSet: number;
    systemFields?: SystemField[];
}

export interface DataClassModelConfig extends DataClassOptionsConfig {
    domain: DomainDesign;
    exception: string;
}

export class DataClassModel implements DataClassModelConfig {
    [immerable] = true;

    readonly category: string;
    readonly description: string;
    readonly domain: DomainDesign;
    readonly exception: string;
    readonly name: string;
    readonly nameExpression: string;
    readonly rowId: number;
    readonly sampleSet: number;
    readonly systemFields: SystemField[];
    readonly parentAliases?: OrderedMap<string, IParentAlias>;
    readonly importAliases?: Map<string, string>;
    readonly isBuiltIn?: boolean;

    constructor(values?: Partial<DataClassModelConfig>) {
        Object.assign(this, values);
    }

    static create(raw: any): DataClassModel {
        let domain = DomainDesign.create({});
        if (raw.domainDesign) {
            domain = DomainDesign.create(raw.domainDesign);
        }

        const model = new DataClassModel({ ...(raw.options ? raw.options : raw), domain });

        return produce(model, (draft: Draft<DataClassModel>) => {
            if (raw.options && model.category === null) {
                draft.category = undefined;
            }
            if (raw.options && model.sampleSet === null) {
                draft.sampleSet = undefined;
            }

            const aliases = raw.options?.importAliases || {};
            draft.importAliases = Map<string, string>(fromJS(aliases));

            draft.systemFields =
                model.category === 'sources' ? SOURCE_DOMAIN_SYSTEM_FIELDS : DATACLASS_DOMAIN_SYSTEM_FIELDS;
        });
    }

    get containerPath(): string {
        return this.domain?.container;
    }

    get isNew(): boolean {
        return !this.rowId;
    }

    isValid(defaultNameFieldConfig?: Partial<IDomainField>): boolean {
        return (
            this.hasValidProperties &&
            !this.hasInvalidNameField(defaultNameFieldConfig) &&
            getDuplicateAlias(this.parentAliases, true).size === 0 &&
            !this.domain.hasInvalidFields()
        );
    }

    get hasValidProperties(): boolean {
        const hasInvalidAliases =
            this.parentAliases &&
            this.parentAliases.size > 0 &&
            this.parentAliases.find(parentAliasInvalid) !== undefined;

        return this.name !== undefined && this.name !== null && this.name.trim().length > 0 && !hasInvalidAliases;
    }

    hasInvalidNameField(defaultNameFieldConfig: Partial<IDomainField>): boolean {
        return this.domain && defaultNameFieldConfig ? this.domain.hasInvalidNameField(defaultNameFieldConfig) : false;
    }

    get entityDataMap(): Map<string, any> {
        return fromJS({
            rowId: this.rowId,
            name: this.name,
            description: this.description,
            nameExpression: this.nameExpression,
        });
    }

    get options(): DataClassOptionsConfig {
        return {
            rowId: this.rowId,
            name: this.name,
            description: this.description,
            nameExpression: this.nameExpression,
            category: this.category,
            sampleSet: this.sampleSet,
            systemFields: this.systemFields,
        };
    }

    mutate(props: Partial<DataClassModel>): DataClassModel {
        return produce<DataClassModel>(this, draft => {
            Object.assign(draft, props);
        });
    }
}
