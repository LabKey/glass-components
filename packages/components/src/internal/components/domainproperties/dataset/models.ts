/*
 * Copyright (c) 2020 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Record } from 'immutable';

import { immerable, produce } from 'immer';

import { DomainDesign, DomainField } from '../models';

import { DOMAIN_FIELD_FULLY_LOCKED } from '../constants';

import { allowAsManagedField } from './utils';

export interface DatasetAdvancedSettingsForm {
    cohortId?: number;
    datasetId?: number;
    showByDefault?: boolean;
    tag?: string;
    visitDatePropertyName?: string;
}

export interface IDatasetModel {
    category?: string;
    cohortId?: number;
    dataSharing?: string;
    datasetId?: number;
    definitionIsShared?: boolean;
    demographicData: boolean;
    description?: string;
    domain: DomainDesign;
    domainId: number;
    entityId?: string;
    exception: string;
    keyPropertyManaged: boolean;
    keyPropertyName?: string;
    tag?: string;
    visitDatePropertyName?: string;
    showByDefault: boolean;
    sourceName?: string;
    sourceType?: string;
    name: string;
    sourceUrl?: string;
    label?: string;
    useTimeKeyField?: boolean;
}

export class DatasetModel implements IDatasetModel {
    [immerable] = true;

    readonly domain: DomainDesign;
    readonly domainId: number;
    readonly exception: string;
    readonly datasetId?: number;
    readonly entityId?: string;
    readonly name: string;
    readonly category?: string;
    readonly visitDatePropertyName?: string;
    readonly keyPropertyName?: string;
    readonly keyPropertyManaged: boolean;
    readonly demographicData: boolean;
    readonly label?: string;
    readonly cohortId?: number;
    readonly tag?: string;
    readonly showByDefault: boolean;
    readonly description?: string;
    readonly dataSharing?: string;
    readonly definitionIsShared?: boolean;
    readonly sourceName?: string;
    readonly sourceUrl?: string;
    readonly sourceType?: string;
    readonly useTimeKeyField?: boolean;

    constructor(datasetModel: IDatasetModel) {
        Object.assign(this, datasetModel);
    }

    static create(newDataset = null, raw?: any): DatasetModel {
        if (newDataset) {
            const domain = DomainDesign.create(undefined);
            return new DatasetModel({ ...newDataset, domain });
        } else {
            const domain = DomainDesign.create(raw.domainDesign);
            let model = new DatasetModel({ ...raw.options, domain });

            // if the dataset is from a linked source, disable/lock the fields
            if (model.isFromLinkedSource()) {
                const newDomain = domain.merge({
                    fields: domain.fields
                        .map((field: DomainField) => {
                            return field.set('lockType', DOMAIN_FIELD_FULLY_LOCKED);
                        })
                        .toList(),
                }) as DomainDesign;

                model = produce<DatasetModel>(model, draft => {
                    draft.domain = newDomain;
                });
            }

            return model;
        }
    }

    hasValidProperties(): boolean {
        const isValidName = this.name !== undefined && this.name !== null && this.name.trim().length > 0;
        const isValidLabel =
            this.isNew() || (this.label !== undefined && this.label !== null && this.label.trim().length > 0);

        return isValidName && isValidLabel && this.hasValidAdditionalKey();
    }

    hasValidAdditionalKey(): boolean {
        let isValidKeySetting = true;
        if (this.getDataRowSetting() === 2) {
            isValidKeySetting =
                (this.keyPropertyName !== undefined && this.keyPropertyName !== '') || this.useTimeKeyField;
        }

        return isValidKeySetting;
    }

    isNew(): boolean {
        return !this.entityId;
    }

    getDataRowSetting(): number {
        const noKeyPropName = this.keyPropertyName === undefined || this.keyPropertyName === null;

        // participant id
        if (noKeyPropName && this.demographicData) {
            return 0;
        }
        // participant id and timepoint
        else if (noKeyPropName && !this.useTimeKeyField) {
            return 1;
        }

        // participant id, timepoint and additional key field
        return 2;
    }

    validManagedKeyField(editedName?: string): boolean {
        if (this.keyPropertyName || editedName) {
            const domainFields = this.domain.fields;

            const allowedFieldTypes = domainFields
                .filter(field => allowAsManagedField(field))
                .map(field => {
                    return field.name;
                })
                .toList();

            return allowedFieldTypes.contains(this.keyPropertyName) || allowedFieldTypes.contains(editedName);
        } else {
            return false;
        }
    }

    getDomainKind(timepointType: string): string {
        return timepointType === 'VISIT' ? 'StudyDatasetVisit' : 'StudyDatasetDate';
    }

    getOptions(): Record<string, any> {
        return produce<IDatasetModel>(this, draft => {
            delete draft.exception;
            delete draft.domain;
        });
    }

    isValid(): boolean {
        return this.hasValidProperties() && !this.domain.hasInvalidFields();
    }

    isFromLinkedSource(): boolean {
        return this.sourceName !== undefined && this.sourceName !== null;
    }
}
