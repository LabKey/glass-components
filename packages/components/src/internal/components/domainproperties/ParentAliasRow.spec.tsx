import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { SELECT_INPUT_CONTROL_SELECTOR } from '../forms/input/SelectInputTestUtils';

import {IParentAlias, IParentOption} from '../entities/models';

import { ParentAliasRow } from './ParentAliasRow';

describe('<ParentAliasRow/>', () => {
    test('Ignore empty values', () => {
        const parentAlias: IParentAlias = {
            alias: '',
            id: '',
            parentValue: undefined,
            ignoreAliasError: true,
            ignoreSelectError: true,
        };

        const component = (
            <ParentAliasRow
                id={parentAlias.id}
                onAliasChange={jest.fn()}
                onRemove={jest.fn()}
                parentAlias={parentAlias}
                parentOptions={[]}
            />
        );

        const wrapper = mount(component);

        expect(wrapper.find('input[name="alias"]').props().value).toBe(parentAlias.alias);
        expect(wrapper.find('.has-error')).toHaveLength(0);
    });

    test('Blank values, Error CSS applied', () => {
        const parentAlias: IParentAlias = {
            alias: '',
            id: 'testId',
            parentValue: undefined,
            ignoreAliasError: false,
            ignoreSelectError: false,
        };

        const component = (
            <ParentAliasRow
                id={parentAlias.id}
                onAliasChange={jest.fn()}
                onRemove={jest.fn()}
                parentAlias={parentAlias}
                parentOptions={[]}
            />
        );

        const wrapper = mount(component);
        const aliasInput = wrapper.find('input[name="alias"]');

        expect(wrapper.find('.has-error input[name="alias"]')).toHaveLength(1);
        expect(aliasInput.props().value).toBe(parentAlias.alias);
        expect(wrapper.find(`.has-error ${SELECT_INPUT_CONTROL_SELECTOR}`)).toHaveLength(1);
    });

    test('With values', () => {
        const parentAlias: IParentAlias = {
            alias: 'testAlias',
            id: 'testId',
            ignoreAliasError: false,
            ignoreSelectError: false,
            parentValue: { value: 'materialInputs/test', label: 'Test Label' },
        };
        const option: IParentOption = {
            label: 'test',
            query: 'sampleset',
            schema: 'exp',
            value: 'materialInputs/test',
        };

        const component = (
            <ParentAliasRow
                id={parentAlias.id}
                onAliasChange={jest.fn()}
                onRemove={jest.fn()}
                parentAlias={parentAlias}
                parentOptions={[option]}
            />
        );

        const tree = renderer.create(component);
        expect(tree).toMatchSnapshot();

        const wrapper = mount(component);

        expect(wrapper.find('input[name="alias"]').props().value).toBe(parentAlias.alias);
        expect(wrapper.find('input[name="parentValue"]').props().value).toBe(parentAlias.parentValue.value);
    });

    test('With parent value not an option in select', () => {
        const parentAlias: IParentAlias = {
            alias: 'testAlias',
            id: 'testId',
            ignoreAliasError: false,
            ignoreSelectError: false,
            parentValue: { value: 'materialInputs/test', label: 'Test Label' },
        };
        const option: IParentOption = {
            label: 'test',
            query: 'sampleset',
            schema: 'exp',
            value: 'materialInputs/notFound',
        };

        const component = (
            <ParentAliasRow
                id={parentAlias.id}
                onAliasChange={jest.fn()}
                onRemove={jest.fn()}
                parentAlias={parentAlias}
                parentOptions={[option]}
            />
        );

        const tree = renderer.create(component);
        expect(tree).toMatchSnapshot();

        const wrapper = mount(component);
        expect(wrapper.find('input[name="alias"]').props().value).toBe(parentAlias.alias);
        expect(wrapper.find('input[name="parentValue"]').props().value).toBe(parentAlias.parentValue.value);
    });

    test('Simulate delete behavior', () => {
        const parentAlias: IParentAlias = {
            alias: 'testAlias',
            id: 'testId',
            ignoreAliasError: false,
            ignoreSelectError: false,
            parentValue: { value: 'materialInputs/option2', label: 'Test Label' },
        };
        const option1: IParentOption = {
            label: 'option1',
            query: 'sampleset',
            schema: 'exp',
            value: 'materialInputs/option1',
        };
        const option2: IParentOption = {
            label: 'option2',
            query: 'sampleset',
            schema: 'exp',
            value: 'materialInputs/option2',
        };

        const mockRemove = jest.fn();

        const component = (
            <ParentAliasRow
                id={parentAlias.id}
                onAliasChange={jest.fn()}
                onRemove={mockRemove}
                parentAlias={parentAlias}
                parentOptions={[option1, option2]}
            />
        );

        const wrapper = mount(component);
        const removeBtn = wrapper.find('i.container--removal-icon');
        expect(mockRemove).toHaveBeenCalledTimes(0);
        removeBtn.simulate('click');
        expect(mockRemove).toHaveBeenCalledTimes(1);

        // Remove should be called with the ID to be removed
        expect(mockRemove).toHaveBeenCalledWith(parentAlias.id);
    });

    test('Simulate changed alias', () => {
        const parentAlias: IParentAlias = {
            alias: 'testAlias',
            id: 'testId',
            parentValue: { value: 'materialInputs/option2', label: 'Test Label' },
            ignoreAliasError: false,
            ignoreSelectError: false,
        };
        const option1: IParentOption = {
            label: 'option1',
            query: 'sampleset',
            schema: 'exp',
            value: 'materialInputs/option1',
        };
        const option2: IParentOption = {
            label: 'option2',
            query: 'sampleset',
            schema: 'exp',
            value: 'materialInputs/option2',
        };

        const mockChange = jest.fn();

        const component = (
            <ParentAliasRow
                id={parentAlias.id}
                onAliasChange={mockChange}
                onRemove={jest.fn()}
                parentAlias={parentAlias}
                parentOptions={[option1, option2]}
            />
        );

        const wrapper = mount(component);
        const aliasInput = wrapper.find('input[name="alias"]');
        expect(aliasInput.props().value).toBe(parentAlias.alias);
        expect(mockChange).toHaveBeenCalledTimes(0);

        aliasInput.simulate('change', { target: { name: 'alias', value: 'change' } });
        expect(mockChange).toHaveBeenCalledTimes(1);
        expect(mockChange).toBeCalledWith('testId', 'alias', 'change');
        mockChange.mockClear();
    });
});
