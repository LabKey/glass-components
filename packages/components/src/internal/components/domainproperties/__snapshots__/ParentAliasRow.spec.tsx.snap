// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ParentAliasRow/> With parent value not an option in select 1`] = `
<div
  className="row"
>
  <div
    className="col-xs-2"
  >
    Parent 
    <span
      className="domain-no-wrap"
    >
      Alias
      <div
        className="overlay-trigger"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <span
          className="label-help-target"
        >
          <span
            className="label-help-icon fa fa-question-circle"
          />
        </span>
      </div>
       *
    </span>
  </div>
  <div
    className="col-xs-3"
  >
    <input
      className="form-control"
      defaultValue="testAlias"
      name="alias"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Enter a parent alias for import"
      type="text"
    />
  </div>
  <div
    className="col-xs-5"
  >
    <div
      className="select-input-container form-group row"
    >
      <div
        className="import-alias--parent-select"
      >
        <div
          className="select-input css-b62m3t-container"
          id="selectinput-9"
          onKeyDown={[Function]}
        >
          <span
            className="css-1f43avz-a11yText-A11yText"
            id="react-select-6-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            className="css-1f43avz-a11yText-A11yText"
            role="log"
          />
          <div
            className="select-input__control css-13cymwt-control"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className="select-input__value-container select-input__value-container--has-value css-hlgwow"
            >
              <div
                className="select-input__single-value css-1dimb5e-singleValue"
              >
                materialInputs/test
              </div>
              <div
                className="select-input__input-container css-19bb58m"
                data-value=""
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-expanded={false}
                  aria-haspopup={true}
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  className="select-input__input"
                  disabled={false}
                  id="react-select-6-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  role="combobox"
                  spellCheck="false"
                  style={
                    {
                      "background": 0,
                      "border": 0,
                      "color": "inherit",
                      "font": "inherit",
                      "gridArea": "1 / 2",
                      "label": "input",
                      "margin": 0,
                      "minWidth": "2px",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "100%",
                    }
                  }
                  tabIndex={0}
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              className="select-input__indicators css-1wy0on6"
            >
              <div
                aria-hidden="true"
                className="select-input__indicator select-input__clear-indicator css-1xc3v61-indicatorContainer"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="css-tj5bde-Svg"
                  focusable="false"
                  height={20}
                  viewBox="0 0 20 20"
                  width={20}
                >
                  <path
                    d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                  />
                </svg>
              </div>
              <span
                className="select-input__indicator-separator css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                className="select-input__indicator select-input__dropdown-indicator css-1xc3v61-indicatorContainer"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="css-tj5bde-Svg"
                  focusable="false"
                  height={20}
                  viewBox="0 0 20 20"
                  width={20}
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <input
            name="parentValue"
            type="hidden"
            value="materialInputs/test"
          />
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      className="entity-insert--remove-parent"
    >
      <span
        className="container--action-button"
        onClick={[Function]}
      >
        <i
          className="fa fa-times container--removal-icon"
        />
        
      </span>
    </div>
  </div>
</div>
`;

exports[`<ParentAliasRow/> With values 1`] = `
<div
  className="row"
>
  <div
    className="col-xs-2"
  >
    Parent 
    <span
      className="domain-no-wrap"
    >
      Alias
      <div
        className="overlay-trigger"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <span
          className="label-help-target"
        >
          <span
            className="label-help-icon fa fa-question-circle"
          />
        </span>
      </div>
       *
    </span>
  </div>
  <div
    className="col-xs-3"
  >
    <input
      className="form-control"
      defaultValue="testAlias"
      name="alias"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Enter a parent alias for import"
      type="text"
    />
  </div>
  <div
    className="col-xs-5"
  >
    <div
      className="select-input-container form-group row"
    >
      <div
        className="import-alias--parent-select"
      >
        <div
          className="select-input css-b62m3t-container"
          id="selectinput-5"
          onKeyDown={[Function]}
        >
          <span
            className="css-1f43avz-a11yText-A11yText"
            id="react-select-4-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            className="css-1f43avz-a11yText-A11yText"
            role="log"
          />
          <div
            className="select-input__control css-13cymwt-control"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className="select-input__value-container select-input__value-container--has-value css-hlgwow"
            >
              <div
                className="select-input__single-value css-1dimb5e-singleValue"
              >
                test
              </div>
              <div
                className="select-input__input-container css-19bb58m"
                data-value=""
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-expanded={false}
                  aria-haspopup={true}
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  className="select-input__input"
                  disabled={false}
                  id="react-select-4-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  role="combobox"
                  spellCheck="false"
                  style={
                    {
                      "background": 0,
                      "border": 0,
                      "color": "inherit",
                      "font": "inherit",
                      "gridArea": "1 / 2",
                      "label": "input",
                      "margin": 0,
                      "minWidth": "2px",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "100%",
                    }
                  }
                  tabIndex={0}
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              className="select-input__indicators css-1wy0on6"
            >
              <div
                aria-hidden="true"
                className="select-input__indicator select-input__clear-indicator css-1xc3v61-indicatorContainer"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="css-tj5bde-Svg"
                  focusable="false"
                  height={20}
                  viewBox="0 0 20 20"
                  width={20}
                >
                  <path
                    d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                  />
                </svg>
              </div>
              <span
                className="select-input__indicator-separator css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                className="select-input__indicator select-input__dropdown-indicator css-1xc3v61-indicatorContainer"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="css-tj5bde-Svg"
                  focusable="false"
                  height={20}
                  viewBox="0 0 20 20"
                  width={20}
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <input
            name="parentValue"
            type="hidden"
            value="materialInputs/test"
          />
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      className="entity-insert--remove-parent"
    >
      <span
        className="container--action-button"
        onClick={[Function]}
      >
        <i
          className="fa fa-times container--removal-icon"
        />
        
      </span>
    </div>
  </div>
</div>
`;
